File: partiql-plan/src/main/java/org/partiql/plan/rex/RexLit.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operator;
import org.partiql.plan.OperatorVisitor;
import org.partiql.spi.value.Datum;

import java.util.List;

/**
 * Literal value expression abstract base class.
 */
public abstract class RexLit extends RexBase {

    /**
     * @return new RexLit instance
     */
    @NotNull
    public static RexLit create(@NotNull Datum value) {
        return new Impl(value);
    }

    @NotNull
    public abstract Datum getDatum();

    @NotNull
    @Override
    protected final RexType type() {
        return RexType.of(getDatum().getType());
    }

    @Override
    protected List<Operator> operands() {
        return List.of();
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitLit(this, ctx);
    }

    private static class Impl extends RexLit {

        private final Datum value;

        private Impl(Datum value) {
            this.value = value;
        }

        @NotNull
        @Override
        public Datum getDatum() {
            return value;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexLit.kt
package org.partiql.plan.rex

import org.partiql.spi.value.Datum

/**
 * TODO DOCUMENTATION
 */
public interface RexLit : Rex {

    public fun getValue(): Datum

    override fun getChildren(): Collection<Rex> = emptyList()

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitLit(this, ctx)
}

internal class RexLitImpl(value: Datum) : RexLit {

    // DO NOT USE FINAL
    private var _value = value

    override fun getValue(): Datum = _value

    override fun getType(): RexType = RexType.of(_value.type)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexLit) return false
        if (_value != other.getValue()) return false
        return true
    }

    override fun hashCode(): Int = _value.hashCode()
}
